evalml/pipelines/component_graph.py

```diff
@@ -0,0 +1,295 @@
+from evalml.pipelines.components import ComponentBase
+from evalml.pipelines.components.utils import handle_component_class
+from evalml.utils import import_or_raise
+
+
+class ComponentGraph:
+    def __init__(self, component_dict=None, random_state=0):
+        """ Initializes a component graph for a pipeline as a DAG.
+        """
+        self.component_dict = component_dict or {}
+        for key, value in self.component_dict.items():
+            if not isinstance(value, list):
+                self.component_dict[key] = [value]
+        self._compute_order = []
+        self._recompute_order()
+        self.random_state = random_state
+
+    def from_list(self, component_list):
+        """Constructs a linear graph from a given list
+
+        Arguments:
+            component_list (list): String names or ComponentBase subclasses in
+                                   an order that represents a valid linear graph
+        """
+        for idx, component in enumerate(component_list):
+            component_class = handle_component_class(component)
+            component_name = component_class.name
+
+            parent = None
+            if idx != 0:
+                parent = [f"{handle_component_class(component_list[idx - 1]).name}.x"]
+            self.add_node(component_name, component_class, parents=parent)
+        return self
+
+    def instantiate(self, parameters):
+        """Instantiates all uninstantiated components within the graph using the given parameters. An error will be
+        raised if a component is already instantiated but the parameters dict contains arguments for that component.
+
+        Arguments:
+            parameters (dict): Dictionary with component names as keys and dictionary of that component's parameters as values.
+                               An empty dictionary {} implies using all default values for component parameters.
+        """
+        for component_name, component_info in self.component_dict.items():
+            component_parameters = parameters.get(component_name, {})
+            component_class = component_info[0]
+            if isinstance(component_class, ComponentBase):
+                if component_parameters != {}:
+                    raise ValueError(f"Attempting to instantiate component {component_name} with parameters {component_parameters} when component already instantiated")
+                else:
+                    continue
+
+            component_class = handle_component_class(component_class)
+
+            try:
+                new_component = component_class(**component_parameters, random_state=self.random_state)
+            except (ValueError, TypeError) as e:
+                err = "Error received when instantiating component {} with the following arguments {}".format(component_name, component_parameters)
+                raise ValueError(err) from e
+
+            self.component_dict[component_name][0] = new_component
```

Comment:
Interesting idea! Would this `component_instances` dict also contain the edges, or just be a name->instance mapping? Based on your other comments it seems like you'd prefer it also contain the edges, but there I worry about storing repeated information unnecessarily.

Reply:
Hmm, good point! Come to think of it, I guess all you really need to store is a dict mapping component ID to component instance. What do you think of that?

Reply:
A dictionary containing just component names to instances sounds a lot like my original implementation of this class 😅 which I'm totally fine with going back to. The only question there is where the edges should get stored in that case: would these stay in the original user-inputted `component_dict` or also get their own variable? Both choices here seem somewhat inefficient.

It may be important to bring up here: the entire `ComponentGraph` class is going to be private, none of this will be exposed to users in any way.

Reply:
Haha nice.

I believe there's two main things we need to keep track of in `ComponentGraph`:
1. The graph structure: what are the nodes (component classes) and what are the edges connecting the nodes
2. The component instances: create with parameters, fit on training data, transform/predict on other data

I had commented here originally because I think we should use a separate structure to keep item 1 (graph structure) vs item 2 (component instances).

I think its fine to store the component instances as a list, as a dict mapping component name to instance, or in a dict with edges just like the `component_graph` structure, as long as we have that separation.

Personally, I'd find the implementation most intuitive if the component instances lived in a dict mapping component name to component instance. If that's what you had originally, then great let's bring it back! :)

And yep I definitely agree, this won't be exposed to users directly, so we have future-flexibility in our choice.

What do you think?

Reply:
Discussed with @eccabay . Plan:
* On pipelines, keep `component_graph` as user input.
* Create private instance of `ComponentGraph` which holds instances and computes.
* So for this PR, the answer to the question above doesn't matter in terms of what users will interact with